File: src/main/java/edu/cornell/cs/apl/viaduct/security/TrustLattice.java
package edu.cornell.cs.apl.viaduct.security;

/** A lattice whose elements are interpreted as principals. */
public interface TrustLattice<T extends TrustLattice<T>> {
  /** Decide if {@code this} is trusted to enforce {@code that}'s security policies. */
  boolean actsFor(T that);

  /**
   * The least powerful principal that can act for both {@code this} and {@code that}. This denotes
   * a conjunction of authority.
   */
  T and(T that);

  /**
   * The most powerful principal both {@code this} and {@code that} can act for. This denotes a
   * disjunction of authority.
   */
  T or(T that);
}

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/security/TrustLattice.kt
package edu.cornell.cs.apl.viaduct.security

/** A lattice whose elements are interpreted as principals. */
interface TrustLattice<T : TrustLattice<T>?> {
    /** Decide if `this` is trusted to enforce [that]'s security policies. */
    fun actsFor(that: T): Boolean

    /**
     * The least powerful principal that can act for both `this` and [that].
     * This denotes a conjunction of authority.
     */
    infix fun and(that: T): T

    /**
     * The most powerful principal both `this` and [that] can act for.
     * This denotes a disjunction of authority.
     */
    infix fun or(that: T): T
}
